{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TCGA-BRCA Data Loading\n",
    "\n",
    "This notebook loads TCGA-BRCA RNA-seq and clinical data from the GDC API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "\n",
    "# Add src to path\n",
    "sys.path.insert(0, str(Path().resolve().parent / \"src\"))\n",
    "\n",
    "from data_loader import TCGADataLoader"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialize Data Loader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize loader\n",
    "loader = TCGADataLoader(data_dir=\"../data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load File Manifest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Querying GDC API for Gene Expression Quantification files...\n",
      "Found 1231 files\n",
      "Saved manifest to ..\\data\\file_manifest.tsv\n",
      "Found 1231 RNA-seq files\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_id</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_size</th>\n",
       "      <th>case_id</th>\n",
       "      <th>sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>744a6d3d-b666-49aa-8d26-47f34e3d1eb5</td>\n",
       "      <td>94027f46-390c-4dda-ab89-cb1ac0a291cd.rna_seq.a...</td>\n",
       "      <td>4231952</td>\n",
       "      <td>878f975b-94fd-4d69-b7e7-1ed3ac2ee438</td>\n",
       "      <td>9321161b-d568-4d38-aed3-956786a06329</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4ecc1f1a-8ff4-4552-a5e8-7a9652b6d1d5</td>\n",
       "      <td>df45fb41-4511-4dab-b865-fcdfcda0400a.rna_seq.a...</td>\n",
       "      <td>4238832</td>\n",
       "      <td>e4fc0909-f284-4471-866d-d8967b6adcbc</td>\n",
       "      <td>58bfe278-a80f-4722-b286-4d966214d244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1ace2a0c-773d-45b5-8fd6-968c88731bbb</td>\n",
       "      <td>c33ecc28-d5ba-4416-b93d-445b7883b6e8.rna_seq.a...</td>\n",
       "      <td>4239626</td>\n",
       "      <td>e5aae05a-478e-4a55-a27c-12b2b4be302a</td>\n",
       "      <td>1d8f72b7-8596-40e8-afdf-244dd897758c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2d5b0962-b5c4-4482-9f28-47e4dcdb6df6</td>\n",
       "      <td>602e3d24-a45d-4fe4-9b70-a679d93f470d.rna_seq.a...</td>\n",
       "      <td>4221019</td>\n",
       "      <td>16fc3677-0393-4ed1-ad3f-c8355f056369</td>\n",
       "      <td>a31dec9b-3626-4cb0-8265-1425d5ce1c1e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>84cff855-d17a-4c09-b432-53893ae69c1e</td>\n",
       "      <td>03d891b3-8faf-4384-94ce-2015f1ca5df0.rna_seq.a...</td>\n",
       "      <td>4217372</td>\n",
       "      <td>33fe5833-d7a0-43d7-a03d-56985ea448a6</td>\n",
       "      <td>2a679581-38ad-4742-8d7b-89015df11e7d</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                file_id  \\\n",
       "0  744a6d3d-b666-49aa-8d26-47f34e3d1eb5   \n",
       "1  4ecc1f1a-8ff4-4552-a5e8-7a9652b6d1d5   \n",
       "2  1ace2a0c-773d-45b5-8fd6-968c88731bbb   \n",
       "3  2d5b0962-b5c4-4482-9f28-47e4dcdb6df6   \n",
       "4  84cff855-d17a-4c09-b432-53893ae69c1e   \n",
       "\n",
       "                                           file_name  file_size  \\\n",
       "0  94027f46-390c-4dda-ab89-cb1ac0a291cd.rna_seq.a...    4231952   \n",
       "1  df45fb41-4511-4dab-b865-fcdfcda0400a.rna_seq.a...    4238832   \n",
       "2  c33ecc28-d5ba-4416-b93d-445b7883b6e8.rna_seq.a...    4239626   \n",
       "3  602e3d24-a45d-4fe4-9b70-a679d93f470d.rna_seq.a...    4221019   \n",
       "4  03d891b3-8faf-4384-94ce-2015f1ca5df0.rna_seq.a...    4217372   \n",
       "\n",
       "                                case_id                             sample_id  \n",
       "0  878f975b-94fd-4d69-b7e7-1ed3ac2ee438  9321161b-d568-4d38-aed3-956786a06329  \n",
       "1  e4fc0909-f284-4471-866d-d8967b6adcbc  58bfe278-a80f-4722-b286-4d966214d244  \n",
       "2  e5aae05a-478e-4a55-a27c-12b2b4be302a  1d8f72b7-8596-40e8-afdf-244dd897758c  \n",
       "3  16fc3677-0393-4ed1-ad3f-c8355f056369  a31dec9b-3626-4cb0-8265-1425d5ce1c1e  \n",
       "4  33fe5833-d7a0-43d7-a03d-56985ea448a6  2a679581-38ad-4742-8d7b-89015df11e7d  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get file manifest for RNA-seq expression data\n",
    "manifest = loader.get_file_manifest()\n",
    "print(f\"Found {len(manifest)} RNA-seq files\")\n",
    "manifest.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Clinical Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Querying GDC API for clinical data...\n",
      "Found 1098 cases\n",
      "Saved clinical data to ..\\data\\clinical_data.tsv\n",
      "Found 1098 cases\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>age_at_index</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>tumor_stage</th>\n",
       "      <th>ajcc_pathologic_t</th>\n",
       "      <th>ajcc_pathologic_n</th>\n",
       "      <th>ajcc_pathologic_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>e3935ce4-64d3-4a66-ba11-d308b844b410</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>e3b555aa-7f0a-49c6-9b13-182c61a144c1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17ca61a2-607a-45ff-88fa-ef72e80bf891</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7d681cc6-689d-41c8-9e84-e13733089ec9</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>17f275c1-a0d4-487d-8f02-ea279584b4cd</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id age_at_index gender  race  \\\n",
       "0  e3935ce4-64d3-4a66-ba11-d308b844b410         None   None  None   \n",
       "1  e3b555aa-7f0a-49c6-9b13-182c61a144c1         None   None  None   \n",
       "2  17ca61a2-607a-45ff-88fa-ef72e80bf891         None   None  None   \n",
       "3  7d681cc6-689d-41c8-9e84-e13733089ec9         None   None  None   \n",
       "4  17f275c1-a0d4-487d-8f02-ea279584b4cd         None   None  None   \n",
       "\n",
       "  vital_status days_to_death days_to_birth primary_diagnosis tumor_stage  \\\n",
       "0         None          None          None              None        None   \n",
       "1         None          None          None              None        None   \n",
       "2         None          None          None              None        None   \n",
       "3         None          None          None              None        None   \n",
       "4         None          None          None              None        None   \n",
       "\n",
       "  ajcc_pathologic_t ajcc_pathologic_n ajcc_pathologic_m  \n",
       "0              None              None              None  \n",
       "1              None              None              None  \n",
       "2              None              None              None  \n",
       "3              None              None              None  \n",
       "4              None              None              None  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get clinical data\n",
    "clinical = loader.get_clinical_data()\n",
    "print(f\"Found {len(clinical)} cases\")\n",
    "clinical.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load PAM50 Subtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Querying GDC API for PAM50 subtype annotations...\n",
      "  Method 1: Checking annotations endpoint...\n",
      "  Method 2: Searching for supplementary files with PAM50...\n",
      "  Method 3: Checking clinical data for subtype fields...\n",
      "\n",
      "  Method 4: Attempting automatic download from TCGA publications...\n",
      "PAM50 file already exists: ..\\data\\BRCA.547.PAM50.SigClust.Subtypes.txt\n",
      "Loading PAM50 subtypes from file: ..\\data\\BRCA.547.PAM50.SigClust.Subtypes.txt\n",
      "  Loaded 547 rows from TSV file\n",
      "  Found 547 samples with PAM50 subtypes\n",
      "Found 547 PAM50 annotations\n",
      "\n",
      "PAM50 Subtype Distribution:\n",
      "pam50_subtype\n",
      "LumA      232\n",
      "LumB      129\n",
      "Basal      98\n",
      "Her2       58\n",
      "Normal     30\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Get PAM50 subtype annotations (with auto-download if not found)\n",
    "pam50 = loader.get_pam50_subtypes(auto_download=True)\n",
    "if not pam50.empty:\n",
    "    print(f\"Found {len(pam50)} PAM50 annotations\")\n",
    "    print(f\"\\nPAM50 Subtype Distribution:\")\n",
    "    if 'pam50_subtype' in pam50.columns:\n",
    "        print(pam50['pam50_subtype'].value_counts())\n",
    "    pam50.head()\n",
    "else:\n",
    "    print(\"PAM50 subtypes not found. Trying manual download...\")\n",
    "    try:\n",
    "        file_path = loader.download_pam50_from_url()\n",
    "        pam50 = loader.load_pam50_from_file(str(file_path))\n",
    "        print(f\"Successfully loaded {len(pam50)} PAM50 annotations\")\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to download PAM50: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Summary statistics\n",
    "print(\"=\" * 60)\n",
    "print(\"DATA SUMMARY\")\n",
    "print(\"=\" * 60)\n",
    "\n",
    "print(f\"\\nFile Manifest: {len(manifest)} files\")\n",
    "print(f\"Clinical Data: {len(clinical)} cases\")\n",
    "if not pam50.empty:\n",
    "    print(f\"PAM50 Annotations: {len(pam50)} cases\")\n",
    "\n",
    "print(\"\\nClinical Data Columns:\")\n",
    "for col in clinical.columns:\n",
    "    print(f\"  - {col}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Cleaning & Preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check data types and basic info\n",
    "print(\"Clinical Data Info:\")\n",
    "print(clinical.info())\n",
    "print(\"\\n\" + \"=\"*60)\n",
    "print(\"\\nManifest Info:\")\n",
    "print(manifest.info())\n",
    "print(\"\\n\" + \"=\"*60)\n",
    "if not pam50.empty:\n",
    "    print(\"\\nPAM50 Info:\")\n",
    "    print(pam50.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for missing values across all datasets\n",
    "print(\"Missing Values in Clinical Data:\")\n",
    "missing_clinical = clinical.isnull().sum()\n",
    "missing_clinical = missing_clinical[missing_clinical > 0].sort_values(ascending=False)\n",
    "if len(missing_clinical) > 0:\n",
    "    print(missing_clinical)\n",
    "    print(f\"\\nTotal missing values: {missing_clinical.sum()}\")\n",
    "    print(f\"Percentage missing: {(missing_clinical.sum() / clinical.size) * 100:.2f}%\")\n",
    "else:\n",
    "    print(\"No missing values found!\")\n",
    "\n",
    "print(\"\\n\" + \"=\"*60)\n",
    "print(\"Missing Values in Manifest:\")\n",
    "missing_manifest = manifest.isnull().sum()\n",
    "missing_manifest = missing_manifest[missing_manifest > 0].sort_values(ascending=False)\n",
    "if len(missing_manifest) > 0:\n",
    "    print(missing_manifest)\n",
    "else:\n",
    "    print(\"No missing values found!\")\n",
    "\n",
    "print(\"\\n\" + \"=\"*60)\n",
    "if not pam50.empty:\n",
    "    print(\"Missing Values in PAM50:\")\n",
    "    missing_pam50 = pam50.isnull().sum()\n",
    "    missing_pam50 = missing_pam50[missing_pam50 > 0].sort_values(ascending=False)\n",
    "    if len(missing_pam50) > 0:\n",
    "        print(missing_pam50)\n",
    "    else:\n",
    "        print(\"No missing values found!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean clinical data\n",
    "clinical_clean = clinical.copy()\n",
    "\n",
    "# Convert numeric columns\n",
    "numeric_cols = ['age_at_index', 'days_to_death', 'days_to_birth']\n",
    "for col in numeric_cols:\n",
    "    if col in clinical_clean.columns:\n",
    "        clinical_clean[col] = pd.to_numeric(clinical_clean[col], errors='coerce')\n",
    "\n",
    "# Clean up tumor stage (remove leading text like 'stage ')\n",
    "if 'tumor_stage' in clinical_clean.columns:\n",
    "    clinical_clean['tumor_stage'] = clinical_clean['tumor_stage'].str.replace('stage ', '', case=False, regex=False)\n",
    "\n",
    "# Standardize gender values\n",
    "if 'gender' in clinical_clean.columns:\n",
    "    clinical_clean['gender'] = clinical_clean['gender'].str.strip().str.title()\n",
    "\n",
    "# Create age groups\n",
    "if 'age_at_index' in clinical_clean.columns:\n",
    "    clinical_clean['age_group'] = pd.cut(\n",
    "        clinical_clean['age_at_index'],\n",
    "        bins=[0, 40, 50, 60, 70, 100],\n",
    "        labels=['<40', '40-50', '50-60', '60-70', '70+']\n",
    "    )\n",
    "\n",
    "print(\"Clinical data cleaned!\")\n",
    "print(f\"Original shape: {clinical.shape}\")\n",
    "print(f\"Cleaned shape: {clinical_clean.shape}\")\n",
    "print(f\"\\nFirst few rows:\")\n",
    "clinical_clean.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean PAM50 Subtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean PAM50 subtypes if available\n",
    "if not pam50.empty:\n",
    "    pam50_clean = pam50.copy()\n",
    "    \n",
    "    # Standardize PAM50 subtype names (remove extra spaces, standardize case)\n",
    "    if 'pam50_subtype' in pam50_clean.columns:\n",
    "        pam50_clean['pam50_subtype'] = pam50_clean['pam50_subtype'].str.strip()\n",
    "        # Common PAM50 subtypes: LumA, LumB, Basal, Her2, Normal\n",
    "        # Map variations to standard names\n",
    "        subtype_mapping = {\n",
    "            'Luminal A': 'LumA',\n",
    "            'Luminal B': 'LumB',\n",
    "            'Basal-like': 'Basal',\n",
    "            'HER2-enriched': 'Her2',\n",
    "            'HER2': 'Her2',\n",
    "            'Normal-like': 'Normal'\n",
    "        }\n",
    "        pam50_clean['pam50_subtype'] = pam50_clean['pam50_subtype'].replace(subtype_mapping)\n",
    "        \n",
    "        # Filter to only valid subtypes (drop Unknown/NA)\n",
    "        valid_subtypes = ['LumA', 'LumB', 'Basal', 'Her2', 'Normal']\n",
    "        pam50_clean = pam50_clean[pam50_clean['pam50_subtype'].isin(valid_subtypes)].copy()\n",
    "        \n",
    "        print(f\"PAM50 subtypes cleaned!\")\n",
    "        print(f\"Original count: {len(pam50)}\")\n",
    "        print(f\"Cleaned count: {len(pam50_clean)}\")\n",
    "        print(f\"\\nSubtype distribution:\")\n",
    "        print(pam50_clean['pam50_subtype'].value_counts())\n",
    "        pam50_clean.head()\n",
    "    else:\n",
    "        print(\"No 'pam50_subtype' column found in PAM50 data\")\n",
    "        pam50_clean = pam50.copy()\n",
    "else:\n",
    "    print(\"No PAM50 data available\")\n",
    "    pam50_clean = pd.DataFrame()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge Clinical Data with PAM50 Subtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge clinical data with PAM50 subtypes\n",
    "if not pam50_clean.empty:\n",
    "    # Merge on case_id\n",
    "    clinical_with_subtypes = clinical_clean.merge(\n",
    "        pam50_clean[['case_id', 'pam50_subtype']],\n",
    "        on='case_id',\n",
    "        how='left'\n",
    "    )\n",
    "    \n",
    "    print(f\"Merged dataset shape: {clinical_with_subtypes.shape}\")\n",
    "    print(f\"\\nCases with PAM50 subtypes: {clinical_with_subtypes['pam50_subtype'].notna().sum()}\")\n",
    "    print(f\"Cases without PAM50 subtypes: {clinical_with_subtypes['pam50_subtype'].isna().sum()}\")\n",
    "    print(f\"\\nSubtype distribution:\")\n",
    "    print(clinical_with_subtypes['pam50_subtype'].value_counts(dropna=False))\n",
    "    \n",
    "    clinical_with_subtypes.head()\n",
    "else:\n",
    "    print(\"Cannot merge: PAM50 data not available\")\n",
    "    clinical_with_subtypes = clinical_clean.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Explore Cleaned Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Summary statistics for cleaned clinical data\n",
    "if not clinical_with_subtypes.empty:\n",
    "    print(\"Summary Statistics:\")\n",
    "    print(\"=\"*60)\n",
    "    \n",
    "    # Categorical variables\n",
    "    if 'gender' in clinical_with_subtypes.columns:\n",
    "        print(\"\\nGender Distribution:\")\n",
    "        print(clinical_with_subtypes['gender'].value_counts(dropna=False))\n",
    "    \n",
    "    if 'vital_status' in clinical_with_subtypes.columns:\n",
    "        print(\"\\nVital Status Distribution:\")\n",
    "        print(clinical_with_subtypes['vital_status'].value_counts(dropna=False))\n",
    "    \n",
    "    if 'tumor_stage' in clinical_with_subtypes.columns:\n",
    "        print(\"\\nTumor Stage Distribution:\")\n",
    "        print(clinical_with_subtypes['tumor_stage'].value_counts(dropna=False).head(10))\n",
    "    \n",
    "    if 'age_group' in clinical_with_subtypes.columns:\n",
    "        print(\"\\nAge Group Distribution:\")\n",
    "        print(clinical_with_subtypes['age_group'].value_counts(dropna=False))\n",
    "    \n",
    "    # Numeric variables\n",
    "    print(\"\\nNumeric Statistics:\")\n",
    "    numeric_cols_clean = clinical_with_subtypes.select_dtypes(include=['float64', 'int64']).columns\n",
    "    if len(numeric_cols_clean) > 0:\n",
    "        print(clinical_with_subtypes[numeric_cols_clean].describe())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save Cleaned Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save cleaned datasets\n",
    "data_dir = Path(\"../data\")\n",
    "\n",
    "# Save cleaned clinical data\n",
    "clinical_clean_path = data_dir / \"clinical_data_clean.tsv\"\n",
    "clinical_clean.to_csv(clinical_clean_path, sep=\"\\t\", index=False)\n",
    "print(f\"Saved cleaned clinical data to: {clinical_clean_path}\")\n",
    "\n",
    "# Save clinical data with subtypes if available\n",
    "if not pam50_clean.empty:\n",
    "    clinical_subtypes_path = data_dir / \"clinical_data_with_subtypes.tsv\"\n",
    "    clinical_with_subtypes.to_csv(clinical_subtypes_path, sep=\"\\t\", index=False)\n",
    "    print(f\"Saved clinical data with subtypes to: {clinical_subtypes_path}\")\n",
    "    \n",
    "    # Save cleaned PAM50\n",
    "    pam50_clean_path = data_dir / \"pam50_subtypes_clean.tsv\"\n",
    "    pam50_clean.to_csv(pam50_clean_path, sep=\"\\t\", index=False)\n",
    "    print(f\"Saved cleaned PAM50 subtypes to: {pam50_clean_path}\")\n",
    "\n",
    "print(\"\\n\" + \"=\"*60)\n",
    "print(\"Data cleaning complete!\")\n",
    "print(\"=\"*60)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Next Steps\n",
    "\n",
    "1. ✅ **Data loaded**: Manifest, Clinical, and PAM50 subtypes\n",
    "2. ✅ **Data cleaned**: Missing values, data types, standardization\n",
    "3. ✅ **Data merged**: Clinical data with PAM50 subtypes\n",
    "4. ⏭️ **Next**: Download and process expression data\n",
    "   - Download expression files using `loader.download_file(file_id, file_name)`\n",
    "   - Parse expression data (TSV format)\n",
    "   - Merge expression data with clinical annotations\n",
    "   - Filter for samples with PAM50 subtypes\n",
    "   - Proceed to preprocessing and ML pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean Clinical Data"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "",
   "name": ""
  },
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
